pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+			err := fmt.Errorf("verifyAuthentication: hash algorithm is not supported")
+			return nil, types.ExtStatusAlgoFail, err
+		}
+
+		cert, err := getCloudCert()
+		if err != nil {
+			log.Infof("verifyAuthentication: can not get server cert, %v\n", err)
+			return nil, extraSt, err
+		}
+
+		if sm.Algo == zauth.HashAlgorithm_HASH_SHA256_32bytes &&
+			bytes.Compare(sm.GetSenderCertHash(), cloudCertHash) != 0 ||
+			sm.Algo == zauth.HashAlgorithm_HASH_SHA256_16bytes &&
+				bytes.Compare(sm.GetSenderCertHash(), cloudCertHash[:hashSha256Len16]) != 0 {
+			err := fmt.Errorf("verifyAuthentication: local server cert hash does not match in authen")
+			log.Infof("verifyAuthentication: local server cert hash(%d) does not match in authen (%d) %v, %v",
+				len(cloudCertHash), len(sm.GetSenderCertHash()), cloudCertHash, sm.GetSenderCertHash())
+			return nil, types.ExtStatusCertMiss, err
+		}
+
+		hash := ComputeSha(data)
+		err = verifyAuthSig(sm.GetSignatureHash(), cert, hash)
+		if err != nil {
+			log.Infof("verifyAuthentication: verifyAuthSig error %v\n", err)
+			return nil, types.ExtStatusSignVerifyFail, err
+		}
+		log.Infof("verifyAuthentication: ok\n")
+	}
+	return data, extraSt, nil
+}
+
+func getCloudCert() (*x509.Certificate, error) {
+	var sCert *x509.Certificate
+	if serverLeafCert == nil {
+		certBytes, err := ioutil.ReadFile(types.ServerCertName)
+		if err != nil {
+			log.Infof("getCloudCert: can not read in server cert file, %v\n", err)
+			return sCert, err
+		}
+		block, _ := pem.Decode(certBytes)
+		if block == nil {
+			err := fmt.Errorf("getCloudCert: can not get client Cert")
+			return sCert, err
+		}
+
+		sCert, err = x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			log.Infof("getCloudCert: can not parse cert %v\n", err)
+			return sCert, err
+		}
+
+		// hash verify using PEM bytes from cloud
+		cloudCertHash = ComputeSha(certBytes)
+		serverLeafCert = sCert
+
+	} else {
+		sCert = serverLeafCert
+	}
+
+	return sCert, nil
+}
+
+// verify the signed data with cloud certificate public key
+func verifyAuthSig(signature []byte, cert *x509.Certificate, hash []byte) error {
+
+	log.Debugf("verifyAuthsig sigdata (len %d) %v\n", len(hash), hash)
+
+	switch pub := cert.PublicKey.(type) {
+	case *rsa.PublicKey:
+		err := rsa.VerifyPKCS1v15(pub, crypto.SHA256, hash, signature)
+		if err != nil {
+			return err
+		}
+		log.Infof("verifyAuthSig: verify rsa ok\n")
+	case *ecdsa.PublicKey:
+
+		sigHalflen := len(signature) / 2
+		rbytes := signature[0:sigHalflen]
+		sbytes := signature[sigHalflen:]
+		r := new(big.Int)
+		s := new(big.Int)
+		r.SetBytes(rbytes)
+		s.SetBytes(sbytes)
+		ok := ecdsa.Verify(pub, hash, r, s)
+		if !ok {
+			return errors.New("ecdsa image signature verification failed")
+		}
+		log.Infof("verifyAuthSig: verify ecdsa ok\n")
+	default:
+		log.Debugf("verifyAuthSig: unknown type of public key\n")
+		return errors.New("unknown type of public key")
```

Comment:
Please include the %T type in error string.

Reply:
ok